consensus/scheme/rolldpos/roundcalculator.go

```diff
@@ -217,10 +227,13 @@ func (c *roundCalculator) newRound(
 		if delegates, err = c.Delegates(height); err != nil {
 			return
 		}
+		if executions, err = c.Executions(height); err != nil {
+			return
+		}
 		if roundNum, roundStartTime, err = c.roundInfo(height, blockInterval, now, toleratedOvertime); err != nil {
 			return
 		}
-		if proposer, err = c.calculateProposer(height, roundNum, delegates); err != nil {
+		if proposer, err = c.calculateProposer(height, roundNum, executions); err != nil {
```

Comment:
the process of selecting proposer needs more discussion i think

the set of delegates are static and limited in size (~80 now), and they are almost guaranteed to be online, so calculating a random one from this set makes sense

while the set of executors is expected to be volatile and dynamic in size, so chances are higher the selected executor is offline, if that happens it will hurt the block production rate